numpy/_core/src/multiarray/item_selection.c

```diff
@@ -2724,6 +2724,15 @@ PyArray_CountNonzero(PyArrayObject *self)
             }
         }
         else {
+            /* Special low-overhead version specific to the float types (and some others) */
+            if (PyArray_ISNOTSWAPPED(self)) {
```

Comment:
Just realized (for reasons): Basically every time you check something like this you also need to check whether the array is aligned.  The above code does that, as you can see.

(Please also check in the other PR.  One way to test in CI could be to use a `"?d"` structured dtype, another to just `view` with a byte shift.  The structured dtype just doesn't check if there is a contiguous only special loop.)

Reply:
Do you mean we should replace `PyArray_ISNOTSWAPPED(self)` with `PyArray_ISALIGNED(self) && PyArray_ISNOTSWAPPED(self)`? What would be the reason for this check?

Reply:
Dereferencing an unaligned pointer like `*(float64 *)ptr` will just kill with a SIGBUS or something on come hardware (it may be slow on others).
It isn't a problem on typical hardware...

Dunno, I think some snaitizer (UBSAN?) would find it, otherwise the typical thing is that debian opens an issue eventually because there is a SIGBUS on sparc or so.